feat: add basic loongarch64 support

Godones opened the pull request:
- Refactor the [repo](https://github.com/aoooos/arceos) code to merge
- Use standard toolchain
- Fix the compile command

![image](https://github.com/rcore-os/arceos/assets/57819926/876ae613-51af-49bd-b8c3-a62f98f0568d)


I will submit other code about interrupt soon.

Godones committed:
feat: add basic loongarch64 support

- Refactor the [repo](https://github.com/aoooos/arceos) code to merge
- Use standard toolchain
- Fix the compile command

Godones committed:
fix: fix the CI error about loongarch64 crate

hky1999 commented on Cargo.lock:
No Cargo.lock!!! 👉https://github.com/rust-lang/cargo/issues/315

hky1999 commented on crates/loongarch64/Cargo.toml:
I really do not think it's a good idea to import loongarch64 as a single crate due to its huge code size.
How about make it a independent repo and import as a dependency?

Godones replied:
Ok, i will do it.

Godones committed:
refactor: use loongarch64 crate

Godones committed:
refactor: Remove most useless code

Godones committed:
feat: pass apps/exception

Godones committed:
feat: Support yield application

1. Use r21 register to record percpu variable address
2. Add page table description, although la64 does not use page table for
   the time being.
3. Modify the kernel page table remap part in runtime. la64 does not
   require remapping.

Godones committed:
Implement support for all apps/task and fs.

This commit includes multiple changes aimed at ensuring all apps/task and fs can be supported properly. These edits affect various files within the 'axhal' module and its subdirectories, as well as specific system configuration aspects outlined in the 'loongarch64-qemu-virt.toml' file.

fix bugs about percpu, use r21 register;
fix bugs about timer interrupt;
Simplify saving and restoring trap context;

Godones committed:
Restore cargo.lock

Godones committed:
feat: Add basic C App support for la64

Since loongarch64 does not seem to have the musl tool chain, here we use
the gnu tool chain.

1. add pointer type definition
2. add float definition
3. add setjmp/longjmp

Godones committed:
feat: support net/display app

Godones commented:
@hky1999 Support for la64 is now complete. Except for the two applications redis and sqlite, which were not tested, other apps have passed. If you have time, please check to see if there are any areas for improvement.

equation314 commented on crates/kernel_guard/src/arch/loongarch64.rs:
```diff
@@ -0,0 +1,19 @@
+#[inline]
+pub fn local_irq_save_and_disable() -> usize {
+    let mut flags: usize = 0;
+    let ie_mask: usize = 1 << 2;
```
This mask can be defined as a global constant.

equation314 commented on crates/kernel_guard/src/arch/loongarch64.rs:
```diff
@@ -0,0 +1,19 @@
+    let ie_mask: usize = 1 << 2;
+    // clear the `IE` bit, and return the old CSR
+    // unsafe { asm!("csrrd {}, 0x0", out(reg) flags) };
+    unsafe { asm!("csrxchg {}, {}, 0x0", inout(reg)flags, in(reg) ie_mask) };
```
Add a space before `flags`

equation314 commented on crates/kernel_guard/src/arch/mod.rs:
```diff
@@ -10,5 +10,8 @@ cfg_if::cfg_if! {
     } else if #[cfg(target_arch = "aarch64")] {
         mod aarch64;
         pub use self::aarch64::*;
+    }else if #[cfg(target_arch = "loongarch64")] {
```
Add a space before `else`

equation314 commented on crates/percpu/src/imp.rs:
```diff
@@ -74,6 +74,8 @@ pub fn get_local_thread_pointer() -> usize {
                 core::arch::asm!("mv {}, gp", out(reg) tp)
             } else if #[cfg(target_arch = "aarch64")] {
                 core::arch::asm!("mrs {}, TPIDR_EL1", out(reg) tp)
+            }else if #[cfg(target_arch = "loongarch64")] {
```
Add a space

equation314 commented on tools/la64/la-qemu.sh:
Could it run on the mainline qemu? So we can remove this script.

Godones replied:
Maybe it can be run on the mainline qemu, but according to existing attempts [qemu-system-loongarch64 v7.1 不能正确启动kernel](https://github.com/foxsen/qemu-loongarch-runenv/issues/4), it is difficult to compile a complete tool chain, so most projects currently use a specific version.

equation314 replied:
What about qemu 8.1.3?

Godones replied:
The main problem is that these versions of qemu require appropriate uefi support. I have not yet tried to compile the latest tool chain. Maybe we can remove the dependency on uefi later, so that there will be no restrictions on the qemu version.

equation314 commented on crates/page_table_entry/src/arch/loongarch64.rs:
```diff
@@ -0,0 +1,148 @@
+        /// Whether the page is not executable.
+        const NX = 1 << 62;
+        /// Whether the privilege Level is restricted. When RPLV is 0, the PTE
+        /// can be accessed by any program with privilege Level highter than PLV.
```
highter -> higher

equation314 commented on crates/page_table_entry/src/arch/loongarch64.rs:
```diff
@@ -0,0 +1,148 @@
+        if !f.contains(PTEFlags::NX) {
+            ret |= Self::EXECUTE;
+        }
+        if f.contains(PTEFlags::PLVL) {
```
`f.contains(PTEFlags::PLVL | PTEFlags::PLVH)`

equation314 commented on crates/page_table_entry/src/arch/loongarch64.rs:
```diff
@@ -0,0 +1,148 @@
+                ret |= Self::USER;
+            }
+        }
+        if !f.contains(PTEFlags::MATL) {
```
`MAT = 0` (SUC) -> `Self::DEVICE`

`MAT = 2` (WUC) -> `Self::UNCACHED`

equation314 replied:
```rust
if f.contains(PTEFlags::MATH) {
  ret |= Self::UNCACHED;
} else if !f.contains(PTEFlags::MATL) {
  ret |= Self::DEVICE;
}
```

And you'd better add a comment to explain it.

equation314 commented on crates/page_table_entry/src/arch/loongarch64.rs:
```diff
@@ -0,0 +1,148 @@
+    fn is_present(&self) -> bool {
+        PTEFlags::from_bits_truncate(self.0 as usize).contains(PTEFlags::V)
+    }
+    fn is_huge(&self) -> bool {
```
Any reasons for not supporting huge pages?

equation314 commented on crates/percpu/src/imp.rs:
```diff
@@ -74,6 +74,8 @@ pub fn get_local_thread_pointer() -> usize {
             } else if #[cfg(target_arch = "aarch64")] {
                 core::arch::asm!("mrs {}, TPIDR_EL1", out(reg) tp)
+            }else if #[cfg(target_arch = "loongarch64")] {
+                core::arch::asm!("move {}, $tp", out(reg) tp)
```
We should use `$r21` for percpu base address.

equation314 commented on modules/axhal/Cargo.toml:
```diff
@@ -54,5 +54,7 @@ arm_gic = { path = "../../crates/arm_gic" }
 dw_apb_uart = { path = "../../crates/dw_apb_uart" }
 
+[target.'cfg(target_arch = "loongarch64")'.dependencies]
+loongarch64 = {git = "https://github.com/Godones/loongArch64",package = "loongArch64" }
```
`package` field is unnecessary. And add a space before `git = `

equation314 replied:
It's not a good idea to use uppercase letters in the crate name, see https://github.com/rust-lang/api-guidelines/discussions/29. Could it be renamed?

equation314 reviewed (changes requested):
Thanks for your contribution! Are you still available? Let’s work on merging this PR.

Godones commented:
> Thanks for your contribution! Are you still available? Let’s work on merging this PR.

I will fix these errors.

Godones committed:
fix: fix some formatting errors

equation314 commented on crates/percpu/src/imp.rs:
```diff
@@ -74,6 +74,8 @@ pub fn get_local_thread_pointer() -> usize {
                 core::arch::asm!("mv {}, gp", out(reg) tp)
             } else if #[cfg(target_arch = "aarch64")] {
                 core::arch::asm!("mrs {}, TPIDR_EL1", out(reg) tp)
+            } else if #[cfg(target_arch = "loongarch64")] {
```
`$21` -> `$r21`

Godones committed:
fix: update page table impl

rename loongArch64 to loongarch64

Godones committed:
Restore cargo.lock on the main branch

Godones committed:
doc: add doc for la64

Remove unnecessary firmware programs and add help documentation.

Godones commented:
@equation314 Do you have time to check and see if there are any other problems?

equation314 commented:
Thanks! Please fix conflicts and CI.

equation314 commented on crates/page_table_entry/src/arch/loongarch64.rs:
```diff
@@ -0,0 +1,161 @@
+        if f.contains(MappingFlags::USER) {
+            ret |= Self::PLVH | Self::PLVL;
+        }
+        if !f.contains(MappingFlags::DEVICE | MappingFlags::UNCACHED) {
```
`!A.contains(B | C)` != `!A.contains(B) && !A.contains(C)`, so this expression is wrong.

You can change it to the following with better readability:

```rust
if !f.contains(MappingFlags::DEVICE) {
  if f.contains(MappingFlags::UNCACHED) {
    // MAT = 2 (WUC)
    ret |= Self::MATH
  } else {
    // MAT = 1 (Coherent Cached)
    ret |= Self::MATL
  }
} // else, MAT = 0 (SUC)
```

Godones replied:
Yes, this is a stupid mistake

equation314 commented on crates/percpu/src/imp.rs:
```diff
@@ -108,6 +110,8 @@ pub fn set_local_thread_pointer(cpu_id: usize) {
                 core::arch::asm!("mv gp, {}", in(reg) tp)
             } else if #[cfg(target_arch = "aarch64")] {
                 core::arch::asm!("msr TPIDR_EL1, {}", in(reg) tp)
+            }else if #[cfg(target_arch = "loongarch64")] {
```
Add space

Godones committed:
fix: fix the condition judgment error

Godones commented:
> Thanks! Please fix conflicts and CI.

@equation314 This CI error may be due to a rust version issue. I'm not very familiar with CI, but it seems that the latest version of rust is used in CI.

Godones committed:
doc: add documentation

hky1999 commented:
> > Thanks! Please fix conflicts and CI.
> 
> @equation314 This CI error may be due to a rust version issue. I'm not very familiar with CI, but it seems that the latest version of rust is used in CI.
 
Hello, CI should fit the latest version of Rust, currently I believe it should be `rust version 1.77.0-nightly (bf8716f1c 2023-12-24)`.

Some warnings like this should be fixed.
```
warning: `axhal` (lib) generated 1 warning
warning: the feature `ip_in_core` has been stable since 1.77.0-nightly and no longer requires an attribute to enable
 --> api/arceos_api/src/lib.rs:6:12
  |
6 | #![feature(ip_in_core)]
  |            ^^^^^^^^^^
  |
  = note: `#[warn(stable_features)]` on by default
```
---------------------------------------------------------------------------------------------------
I try to build this branch tonight.
* BIOS `loongarch_bios_0310.bin`  is downloaded from https://github.com/foxsen/qemu-loongarch-runenv/blob/main/loongarch_bios_0310.bin and put it under `tools/la64`.
     maybe this should be added into .gitignore or some corresponding scripts should be added ?
* QEMU version : 
```
➜  arceos git:(la64) ✗ qemu-system-loongarch64 --version                                                     
QEMU emulator version 8.1.94
Copyright (c) 2003-2023 Fabrice Bellard and the QEMU Project developers
```
* The command is `make A=apps/helloworld ARCH=loongarch64 run`.
* But I just get no output:
```
➜  arceos git:(la64) ✗ make A=apps/helloworld ARCH=loongarch64 run
    Building App: helloworld, Arch: loongarch64, Platform: loongarch64-qemu-virt, App type: rust
cargo build --target loongarch64-unknown-none --target-dir /home/hky/workspace/rcore-os/arceos/target --release  -Z build-std --manifest-path apps/helloworld/Cargo.toml --features "axstd/log-level-warn axstd/bus-pci"
warning: value assigned to `flags` is never read
  --> crates/kernel_guard/src/arch/loongarch64.rs:17:53
   |
17 |     unsafe { asm!("csrxchg {}, {}, 0x0", inout(reg) flags, in(reg) IE_BIT) };
   |                                                     ^^^^^
   |
   = help: maybe it is overwritten before being read?
   = note: `#[warn(unused_assignments)]` on by default

warning: `kernel_guard` (lib) generated 1 warning
warning: the feature `const_maybe_uninit_zeroed` has been stable since 1.75.0 and no longer requires an attribute to enable
  --> modules/axhal/src/lib.rs:30:12
   |
30 | #![feature(const_maybe_uninit_zeroed)]
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(stable_features)]` on by default

warning: `axhal` (lib) generated 1 warning
warning: the feature `ip_in_core` has been stable since 1.77.0-nightly and no longer requires an attribute to enable
 --> api/arceos_api/src/lib.rs:6:12
  |
6 | #![feature(ip_in_core)]
  |            ^^^^^^^^^^
  |
  = note: `#[warn(stable_features)]` on by default

warning: `arceos_api` (lib) generated 1 warning
warning: the feature `ip_in_core` has been stable since 1.77.0-nightly and no longer requires an attribute to enable
  --> ulib/axstd/src/lib.rs:52:12
   |
52 | #![feature(ip_in_core)]
   |            ^^^^^^^^^^
   |
   = note: `#[warn(stable_features)]` on by default

warning: `axstd` (lib) generated 1 warning
    Finished release [optimized] target(s) in 0.11s
rust-objcopy --binary-architecture=loongarch64 apps/helloworld/helloworld_loongarch64-qemu-virt.elf --strip-all -O binary apps/helloworld/helloworld_loongarch64-qemu-virt.bin
    Running on qemu...
qemu-system-loongarch64 -m 1G -smp 1 -bios tools/la64/loongarch_bios_0310.bin -kernel apps/helloworld/helloworld_loongarch64-qemu-virt.elf -nographic -vga none
```
Could you offer some help or advice with it? Thanks a lot! @Godones

Godones commented:
> > > Thanks! Please fix conflicts and CI.
> > 
> > 
> > @equation314 This CI error may be due to a rust version issue. I'm not very familiar with CI, but it seems that the latest version of rust is used in CI.
> 
> Hello, CI should fit the latest version of Rust, currently I believe it should be `rust version 1.77.0-nightly (bf8716f1c 2023-12-24)`.
> 
> Some warnings like this should be fixed.
> 
> ```
> warning: `axhal` (lib) generated 1 warning
> warning: the feature `ip_in_core` has been stable since 1.77.0-nightly and no longer requires an attribute to enable
>  --> api/arceos_api/src/lib.rs:6:12
>   |
> 6 | #![feature(ip_in_core)]
>   |            ^^^^^^^^^^
>   |
>   = note: `#[warn(stable_features)]` on by default
> ```
> 
> I try to build this branch tonight.
> 
> * BIOS `loongarch_bios_0310.bin`  is downloaded from https://github.com/foxsen/qemu-loongarch-runenv/blob/main/loongarch_bios_0310.bin and put it under `tools/la64`.
>   maybe this should be added into .gitignore or some corresponding scripts should be added ?
> * QEMU version :
> 
> ```
> ➜  arceos git:(la64) ✗ qemu-system-loongarch64 --version                                                     
> QEMU emulator version 8.1.94
> Copyright (c) 2003-2023 Fabrice Bellard and the QEMU Project developers
> ```
> 
> * The command is `make A=apps/helloworld ARCH=loongarch64 run`.
> * But I just get no output:
> 
> ```
> ➜  arceos git:(la64) ✗ make A=apps/helloworld ARCH=loongarch64 run
>     Building App: helloworld, Arch: loongarch64, Platform: loongarch64-qemu-virt, App type: rust
> cargo build --target loongarch64-unknown-none --target-dir /home/hky/workspace/rcore-os/arceos/target --release  -Z build-std --manifest-path apps/helloworld/Cargo.toml --features "axstd/log-level-warn axstd/bus-pci"
> warning: value assigned to `flags` is never read
>   --> crates/kernel_guard/src/arch/loongarch64.rs:17:53
>    |
> 17 |     unsafe { asm!("csrxchg {}, {}, 0x0", inout(reg) flags, in(reg) IE_BIT) };
>    |                                                     ^^^^^
>    |
>    = help: maybe it is overwritten before being read?
>    = note: `#[warn(unused_assignments)]` on by default
> 
> warning: `kernel_guard` (lib) generated 1 warning
> warning: the feature `const_maybe_uninit_zeroed` has been stable since 1.75.0 and no longer requires an attribute to enable
>   --> modules/axhal/src/lib.rs:30:12
>    |
> 30 | #![feature(const_maybe_uninit_zeroed)]
>    |            ^^^^^^^^^^^^^^^^^^^^^^^^^
>    |
>    = note: `#[warn(stable_features)]` on by default
> 
> warning: `axhal` (lib) generated 1 warning
> warning: the feature `ip_in_core` has been stable since 1.77.0-nightly and no longer requires an attribute to enable
>  --> api/arceos_api/src/lib.rs:6:12
>   |
> 6 | #![feature(ip_in_core)]
>   |            ^^^^^^^^^^
>   |
>   = note: `#[warn(stable_features)]` on by default
> 
> warning: `arceos_api` (lib) generated 1 warning
> warning: the feature `ip_in_core` has been stable since 1.77.0-nightly and no longer requires an attribute to enable
>   --> ulib/axstd/src/lib.rs:52:12
>    |
> 52 | #![feature(ip_in_core)]
>    |            ^^^^^^^^^^
>    |
>    = note: `#[warn(stable_features)]` on by default
> 
> warning: `axstd` (lib) generated 1 warning
>     Finished release [optimized] target(s) in 0.11s
> rust-objcopy --binary-architecture=loongarch64 apps/helloworld/helloworld_loongarch64-qemu-virt.elf --strip-all -O binary apps/helloworld/helloworld_loongarch64-qemu-virt.bin
>     Running on qemu...
> qemu-system-loongarch64 -m 1G -smp 1 -bios tools/la64/loongarch_bios_0310.bin -kernel apps/helloworld/helloworld_loongarch64-qemu-virt.elf -nographic -vga none
> ```
> 
> Could you offer some help or advice with it? Thanks a lot! @Godones

The Qemu version should be:
```rust
❯ qemu-system-loongarch64 --version
QEMU emulator version 6.2.50 (v6.0.0-7567-gac069a8ffb)
Copyright (c) 2003-2022 Fabrice Bellard and the QEMU Project developers
```
You can complie it according to `tools/la64/la-qemu.sh` or [https://github.com/foxsen/qemu-loongarch-runenv/](https://github.com/foxsen/qemu-loongarch-runenv/)

hky1999 commented on scripts/make/qemu.mk:
```diff
@@ -24,7 +24,17 @@ qemu_args-aarch64 := \
 
+LOONGARCH_BIOS = tools/la64/loongarch_bios_0310.bin
+qemu_args-loongarch64 := \
+  -bios $(LOONGARCH_BIOS) \
```
It seems that it can be boot upon default bios provided by QEMU 8.1.94,
I'll try to verify other applications later, why not just delete this line?
![屏幕截图 2023-12-26 141649](https://github.com/rcore-os/arceos/assets/50011172/a06300c6-a8bd-486a-a159-5b65a8f67e79)

Godones replied:
There should be no bios here.

![image](https://github.com/rcore-os/arceos/assets/57819926/cd8327e3-5767-4e4f-851e-bbd535d7a555)

The first instruction is not located in the BIOS.

Godones replied:
The latest version of la qemu seems to have changed the memory layout, which is causing some loading errors. Also, We need to use the environment prepared by the BIOS for us, such as multi-core startup.
![image](https://github.com/rcore-os/arceos/assets/57819926/16e5edf2-520f-472e-a5ff-5e0188ca1f48)

hky1999 replied:
I see, I tried both SMP=2 and SMP=4 and they got stucked at different place.
Sorry I haven't try your provided QEMU emulator, I appreciate and admire your excellent work.
But @equation314 suggested that code should run on mainstream QEMU version (by the way I personally do not recommend that the system itself should rely a special BIOS version as loongaarch currently provides), maybe we should have a little discussion about it?

equation314 replied:
Is it possible to use mainline QEMU with the specific BIOS?

Godones replied:
I will try it.

Godones replied:
Now, we can use the mainline QEMU for la64. 
I discussed the issue of mainline  qemu with some teachers at loongarch. They said that mainline qemu cannot currently boot the ELF format kernel. because our kernel is not as complex as Linux, we do not need the information passed by the bios, so here I chose to remove the BIOS dependency and start the kernel directly.

Godones replied:
I tested it on qemu-8.2.0. The latest rust tool chain seems to cause several programs under apps/c/pthread to fail.

equation314 replied:
Could you provide a list of the current support status of apps?

Godones replied:
These apps have not been tested: apps/c/iperf|redis|sqlite3. Other apps have been tested.

Godones committed:
feat: remove the bios dep for loongarch64-qemu-virt

Rewrite boot code.
Add pci address space definition
Fix word errors for platforms/*.toml

Godones committed:
remove dead code

equation314 commented on modules/axhal/src/platform/qemu_virt_loongarch64/misc.rs:
```diff
@@ -0,0 +1,9 @@
+/// Shutdown the whole system, including all CPUs.
+pub fn terminate() -> ! {
```
Is there any instruction to exit QEMU? This is necessary for CI.

Godones replied:
I didn't find relevant information on the Internet. Maybe I can ask the engineers at la.

Godones replied:
Mainline qemu does not have an exit function. Is it possible to use other means to exit qemu from the outside?

equation314 commented on tools/la64/README.md:
This doc needs to be updated.

equation314 commented on tools/la64/la-qemu.sh:
This script is no longer in use.

Godones committed:
update doc

AsakuraMizu commented:
Closing due to outdated.
